Objects/unionobject.c

```diff
@@ -0,0 +1,473 @@
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+        PyTypeObject* arg_type =  Py_TYPE(arg);
+        if (arg == Py_None) {
+            arg = (PyObject *)arg_type;
+        }
+        if (PyType_Check(arg)) {
+            if (PyObject_IsInstance(instance, arg) != 0)
+            {
+                return self;
+            }
+        } else if (arg_type == &Py_GenericAliasType) {
+            PyErr_SetString(PyExc_TypeError,
+                    "isinstance() argument 2 cannot contain a parameterized generic");
+            return NULL;
+        }
+    }
+    return instance;
+}
+
+static PyObject *
+union_subclasscheck(PyObject *self, PyObject *instance)
+{
+    if (!PyType_Check(instance)) {
+        PyErr_SetString(PyExc_TypeError, "issubclass() arg 1 must be a class");
+        return NULL;
+    }
+    unionobject *alias = (unionobject *)self;
+    Py_ssize_t nargs = PyTuple_GET_SIZE(alias->args);
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+
+        if (PyType_Check(arg)) {
+            if (PyType_IsSubtype((PyTypeObject *)instance, (PyTypeObject *)arg) != 0)
+            {
+                return self;
+            }
+        } else if (Py_TYPE(arg) == &Py_GenericAliasType) {
+            PyErr_SetString(PyExc_TypeError,
+                    "issubclass() argument 2 cannot contain a parameterized generic");
+            return NULL;
+        }
+    }
+    return instance;
+}
+
+static int
+is_typing_name(PyObject *obj, char *name)
+{
+    PyTypeObject *type = Py_TYPE(obj);
+    if (strcmp(type->tp_name, name) != 0) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)type, "__module__");
+    if (module == NULL) {
+        return -1;
+    }
+    int res = PyUnicode_Check(module)
+        && _PyUnicode_EqualToASCIIString(module, "typing");
+    Py_DECREF(module);
+    return res;
+}
+
+static PyMethodDef union_methods[] = {
+    {"__instancecheck__", union_instancecheck, METH_O},
+    {"__subclasscheck__", union_subclasscheck, METH_O},
+    {0}};
+
+static PyObject *
+union_richcompare(PyObject *a, PyObject *b, int op)
+{
+    if (op != Py_EQ && op != Py_NE) {
+        PyObject *result = Py_NotImplemented;
+        Py_INCREF(result);
+        return result;
+    }
+
+    PyObject* b_set = NULL;
+    unionobject *aa = (unionobject *)a;
+    PyObject* a_set = PySet_New(aa->args);
+    if (a_set == NULL) {
+        Py_RETURN_FALSE;
+    }
+
+    PyTypeObject *type = Py_TYPE(b);
+    if (is_typing_name(b, "_UnionGenericAlias")) {
+        PyObject* b_args = PyObject_GetAttrString(b, "__args__");
+        int b_arg_length = PyTuple_GET_SIZE(b_args);
+        for (int i = 0; i < b_arg_length; i++) {
+            PyObject* arg = PyTuple_GET_ITEM(b_args, i);
+            if (arg == (PyObject *)&_PyNone_Type) {
+                PyTuple_SET_ITEM(b_args, i, Py_None);
```

Comment:
This is due to an issue with `NoneType` vs. `None` in the new union syntax and in `typing.Union`. See this discussion here: https://github.com/python/cpython/pull/21515#discussion_r463390909

If you have any suggestions for how to remedy that situation, I'm all ears :)

Reply:
> If you have any suggestions for how to remedy that situation, I'm all ears :)

Is not enough to do the substitution `None` -> `NoneType` at creation (or union) time? Why do we need to do this lazily?

Reply:
Thanks for clarifying :) Sounds good to me.

Reply:
Sorry, forgot to answer your question here:

> Is not enough to do the substitution None -> NoneType at creation (or union) time? Why do we need to do this lazily? I may be missing something.

We want 

`assert int | None == typing.Union[int, None]`

And the `__repr__` of `int | None` to be `int | None` instead of `int | NoneType`. If we convert it at creation time, we print `NoneType`. As well, the `__args__` should mimic those in the genericalias type:

```
>>> list[None].__args__
(None,)
>>> (int | None).__args__
(<class 'int'>, None)
```


The `None` argument is special cased for `typing.Union` here: https://github.com/python/cpython/blob/master/Lib/typing.py#L394-L395

Reply:
Oh, I understand why we were doing this here. Thanks for the clarification.

> In any case I would prefer to interpose None when doing the comparison instead of mutating the right-hand side object, as comparisons should not have side effects.

In that case, I think it would still be better to substitute the set creation as:
```
        b_set = PySet_New(b_args);
```

by creating an empty set and one by one adding the elements in the tuple, substituting `NoneType` when we encounter `None`.